libnetwork: remove dead / unused code from datastore and kvstore

thaJeztah opened the pull request:
### libnetwork/internal/kvstore: remove unused Config options

The only remaining kvstore is BoltDB, which doesn't use TLS connections
or authentication, so we can remove these options.

### libnetwork/internal/kvstore: fix some linting issues

### libnetwork/internal/kvstore: remove unused WatchTree and NewLock methods

These were not used, and not implemented by the BoltDB store.

### libnetwork/internal/kvstore: remove unused WriteOptions

The WriteOptions struct was only used to set the "IsDir" option. This option
was added in https://github.com/docker/libkv/commit/d635a8e32b67f1ac511b7fcc15d524dedb9f549b (https://github.com/docker/libkv/pull/75)
and was only supported by the etcd libkv store.

The BoltDB store does not support this option, making the WriteOptions
struct fully redundant.

### libnetwork/internal/kvstore: AtomicPut(): remove unused "created" return

This boolean was not used anywhere, so we can remove it.

### libnetwork/internal/kvstore: AtomicDelete(): remove unused "deleted" return

This boolean was not used anywhere, so we can remove it. Also cleaning up
the implementation a bit.

### libnetwork/internal/kvstore/boltdb: minor cleanup/refactor

Make the code slightly more idiomatic; remove some "var" declarations,
remove some intermediate variables and redundant error-checks, and remove
the filePerm const.


### libnetwork/internal/kvstore/boltdb: BoltDB.Exists(): fix error handling

This function could potentially return "true" even if an error was returned.

### libnetwork/internal/kvstore/boltdb: BoltDB.Get(): don't shadow error

Don't shadow the original error if we got one.


### libnetwork/internal/kvstore/boltdb: BoltDB.List(): minor cleanup

cleanup the code to be slightly more idiomatic

### libnetwork/internal/kvstore/boltdb: un-export Mutex

Keep the mutex internal to BoltDB.

### libnetwork/datastore: remove unused datastore.Active()

The value was set, and updated, but never used.

### libnetwork/datastore: remove redundant datastore.sequential

The sequential field determined whether a lock was needed when storing
and retrieving data. This field was always set to true, with the exception
of NewTestDataStore() in the tests.

This field was added in https://github.com/moby/libnetwork/commit/a18e2f9965b0d772ed2b552bb4e26ea43bf90c95 (https://github.com/moby/libnetwork/pull/1206)
to make locking optional for non-local scoped stores. Such stores are no
longer used, so we can remove this field.

### libnetwork/datastore: remove Watch(), Watchable(), RestartWatch()

The `store.Watch()` was only used in `Controller.processEndpointCreate()`,
and skipped if the store was not "watchable" (`store.Watchable()`).

Whether a store is watchable depends on the store's datastore.scope;
local stores are not watchable;

    func (ds *datastore) Watchable() bool {
        return ds.scope != LocalScope
    }

datastore is only initialized in two locations, and both locations set the
scope field to LocalScope:

datastore.newClient() (also called by datastore.NewDataStore()):
https://github.com/moby/moby/blob/3e4c9d90cf6a796334519f0de12ff09608d7abbe/libnetwork/datastore/datastore.go#L213

datastore.NewTestDataStore() (used in tests);
https://github.com/moby/moby/blob/3e4c9d90cf6a796334519f0de12ff09608d7abbe/libnetwork/datastore/datastore_test.go#L14-L17

Furthermore, the backing BoltDB kvstore does not implement the Watch()
method;

https://github.com/moby/moby/blob/3e4c9d90cf6a796334519f0de12ff09608d7abbe/libnetwork/internal/kvstore/boltdb/boltdb.go#L464-L467

Based on the above;

- our datastore is never Watchable()
- so datastore.Watch() is never used

This patch removes the Watchable(), Watch(), and RestartWatch() functions,
as well as the code handling watching.

### libnetwork/internal/kvstore: remove unused Watch() method

The BoltDB store is not Watchable, and the Watch function was never used,
so we can remove it.

### libnetwork/networkdb: NetworkDB.Watch(): remove unused "key" argument

This function was implemented in https://github.com/moby/libnetwork/commit/dd4950f36d0a568791ead9398ddd175451ae36fa
which added a "key" field, but that field was never used anywhere, and
still appears unused.

### libnetwork/datastore: cache.get(): remove unused "key" argument

### libnetwork: rename vars that shadowed with pkg vars and imports

### libnetwork/config: remove options that were only used in tests

The OptionLocalKVProvider, OptionLocalKVProviderURL, and OptionLocalKVProviderConfig
options were only used in tests, so un-export them, and move them to the
test-files.

### libnetwork: inline store config options

### libnetwork: TestBoltdbBackend(): use t.TempDir()

### libnetwork/datastore: remove unused DeleteTree() method


**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah committed:
libnetwork/internal/kvstore: remove unused Config options

The only remaining kvstore is BoltDB, which doesn't use TLS connections
or authentication, so we can remove these options.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork/internal/kvstore: fix some linting issues

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork/internal/kvstore: remove unused WatchTree and NewLock methods

These were not used, and not implemented by the BoltDB store.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork/datastore: remove unused datastore.Active()

The value was set, and updated, but never used.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork/datastore: remove redundant datastore.sequential

The sequential field determined whether a lock was needed when storing
and retrieving data. This field was always set to true, with the exception
of NewTestDataStore() in the tests.

This field was added in https://github.com/moby/libnetwork/commit/a18e2f9965b0d772ed2b552bb4e26ea43bf90c95
to make locking optional for non-local scoped stores. Such stores are no
longer used, so we can remove this field.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork/internal/kvstore: remove unused WriteOptions

The WriteOptions struct was only used to set the "IsDir" option. This option
was added in https://github.com/docker/libkv/commit/d635a8e32b67f1ac511b7fcc15d524dedb9f549b
and was only supported by the etcd libkv store.

The BoltDB store does not support this option, making the WriteOptions
struct fully redundant.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork/datastore: remove Watch(), Watchable(), RestartWatch()

The `store.Watch()` was only used in `Controller.processEndpointCreate()`,
and skipped if the store was not "watchable" (`store.Watchable()`).

Whether a store is watchable depends on the store's datastore.scope;
local stores are not watchable;

    func (ds *datastore) Watchable() bool {
        return ds.scope != LocalScope
    }

datastore is only initialized in two locations, and both locations set the
scope field to LocalScope:

datastore.newClient() (also called by datastore.NewDataStore()):
https://github.com/moby/moby/blob/3e4c9d90cf6a796334519f0de12ff09608d7abbe/libnetwork/datastore/datastore.go#L213

datastore.NewTestDataStore() (used in tests);
https://github.com/moby/moby/blob/3e4c9d90cf6a796334519f0de12ff09608d7abbe/libnetwork/datastore/datastore_test.go#L14-L17

Furthermore, the backing BoltDB kvstore does not implement the Watch()
method;

https://github.com/moby/moby/blob/3e4c9d90cf6a796334519f0de12ff09608d7abbe/libnetwork/internal/kvstore/boltdb/boltdb.go#L464-L467

Based on the above;

- our datastore is never Watchable()
- so datastore.Watch() is never used

This patch removes the Watchable(), Watch(), and RestartWatch() functions,
as well as the code handling watching.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork/internal/kvstore: remove unused Watch() method

The BoltDB store is not Watchable, and the Watch function was never used,
so we can remove it.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork/networkdb: NetworkDB.Watch(): remove unused "key" argument

This function was implemented in https://github.com/moby/libnetwork/commit/dd4950f36d0a568791ead9398ddd175451ae36fa
which added a "key" field, but that field was never used anywhere, and
still appears unused.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork/config: remove options that were only used in tests

The OptionLocalKVProvider, OptionLocalKVProviderURL, and OptionLocalKVProviderConfig
options were only used in tests, so un-export them, and move them to the
test-files.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork: rename vars that shadowed with pkg vars and imports

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork: inline store config options

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork: TestBoltdbBackend(): use t.TempDir()

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork/datastore: remove unused DeleteTree() method

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork/datastore: cache.get(): remove unused "key" argument

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork/internal/kvstore: AtomicPut(): remove unused "created" return

This boolean was not used anywhere, so we can remove it.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork/internal/kvstore/boltdb: minor cleanup/refactor

Make the code slightly more idiomatic; remove some "var" declarations,
remove some intermediate variables and redundant error-checks, and remove
the filePerm const.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork/internal/kvstore: AtomicDelete(): remove unused "deleted" return

This boolean was not used anywhere, so we can remove it. Also cleaning up
the implementation a bit.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork/internal/kvstore/boltdb: BoltDB.List(): minor cleanup

cleanup the code to be slightly more idiomatic

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork/internal/kvstore/boltdb: BoltDB.Exists(): fix error handling

This function could potentially return "true" even if an error was returned.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork/internal/kvstore/boltdb: BoltDB.Get(): don't shadow error

Don't shadow the original error if we got one.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork/internal/kvstore/boltdb: un-export Mutex

Keep the mutex internal to BoltDB.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork/datastore: un-export Mutex

Keep the mutex internal to the DataStore.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork/datastore: remove unused PutObject(), DeleteObject()

all code is using the atomic alternatives for these (PutObjectAtomic,
DeleteObjectAtomic)

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork/internal/kvstore: remove unused Delete()

All code is using the atomic alternatives (AtomicDelete)

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
@akerouanton @corhere @neersighted ptal

corhere commented on libnetwork/datastore/datastore.go:
```diff
@@ -346,10 +345,8 @@ func (ds *datastore) PutObjectAtomic(kvObject KVObject) error {
-		defer ds.Unlock()
-	}
+	ds.Lock()
+	defer ds.Unlock()
```
https://github.com/moby/libnetwork/issues/1201#issuecomment-223185754
> The main purpose of serializing access was for local store where we saw data corruptions.

Sounds like there are some concurrency issues with datastore which the serialization was added to work around. I have a hunch that it involves cache coherency or an impedance mismatch with the libkv API not supporting multi-operation transactions. I hope this cleanup and refactoring work will get us closer to being able to remove the locking.

corhere reviewed (commented):
Almost everything LGTM. I see lots of potential for follow-up work to improve the datastore situation further.

thaJeztah commented on libnetwork/internal/kvstore/boltdb/boltdb.go:
```diff
@@ -24,6 +24,7 @@ var (
 
 // BoltDB type implements the Store interface
 type BoltDB struct {
+	mu         sync.Mutex
```
Ha! I think I ended up on those changes as well.

I have some changes stashed locally; one of those was that I started to remove the `PersistConnection` bool, because it was always `false`;

> libnetwork/kvstore: remove PersistConnection option
>
> The `PersistConnection` option was added to fix issues arround concurrency.
> By default, the boltdb store would use a persistent connection, and keep the
> database opened, which could cause issues if multiple clients tried to re-use
> the connection, and timed out.
> 
> Libkv commit https://github.com/docker/libkv/commit/b291677af7b0b271bd5ab3c876323f0a7028c226
> switched the default to open and close boltDB connnections for every get/pull
> operation, and added the `PersistConnection` option to allow using the prior
> default (keep a persistent connection).
> 
> The change above was pulled into libnetwork in commit:
> https://github.com/moby/libnetwork/commit/3c623fab9fe6cc737f8d1070bb623b08c5277a73
> 
> And the same PR also removed the timeout through commit:
> https://github.com/moby/libnetwork/commits/bc4190db8793f5a4662bb239b77777b78d60772b
> 
> Removing the timeout resulted in performance issues, and the daemon to hang
> (see libnetwork issues 1374, and moby issue 26251), so a change was merged
> in libkv to allow setting a custom timeout, also if `PersistConnection` was
> not used:
> https://github.com/docker/libkv/commit/63d2c8ad0dcfab3f7034530fa75cdbe6270960df
> 
> The current code in this repository never sets the `PersistConnection`, so
> we can remove support for this option.


_HOWEVER_ after I made that change, and wrote the commit message, I started to scratch my head, and wonder ... _WHY_? Why are we not using a persistent connection? Because a quick look at other consumers of boltdb, seem to be using one connection? So yes, I stashed away that commit as I needed to have a closer look / understanding of the situation 😅

thaJeztah commented on libnetwork/internal/kvstore/boltdb/boltdb.go:
```diff
@@ -266,10 +261,13 @@ func (b *BoltDB) List(keyPrefix string) ([]*store.KVPair, error) {
+	if err != nil {
+		return nil, err
+	}
 	if len(kv) == 0 {
 		return nil, store.ErrKeyNotFound
 	}
```
Good point; yes for `List` it doesn't really make sense to treat it as an error; I'll look at it in a follow up 👍

thaJeztah commented:
Okay, I _think_ I addressed your comments for this iteration, but there definitely will be follow-ups. Cutting away some of the cruft at least should help us getting a better view of what's needed, and what's not.

corhere reviewed (approved):
LGTM!

thaJeztah commented:
This one is really flaky (tracked in https://github.com/moby/moby/issues/42358);

```
=== FAIL: arm64.integration.container TestRestartDaemonWithRestartingContainer (1.78s)
[2023-07-05T06:41:27.167Z]     daemon_linux_test.go:211: [d05238003e645] failed to start daemon with arguments [--data-root /go/src/github.com/docker/docker/bundles/test-integration/TestRestartDaemonWithRestartingContainer/d05238003e645/root --exec-root /tmp/dxr/d05238003e645 --pidfile /go/src/github.com/docker/docker/bundles/test-integration/TestRestartDaemonWithRestartingContainer/d05238003e645/docker.pid --userland-proxy=true --containerd-namespace d05238003e645 --containerd-plugins-namespace d05238003e645p --containerd /var/run/docker/containerd/containerd.sock --host unix:///tmp/docker-integration/d05238003e645.sock --debug --storage-driver overlay2] : [d05238003e645] daemon exited during startup: exit status 1
[2023-07-05T06:41:27.167Z]     panic.go:522: [d05238003e645] daemon is not started
```

thaJeztah commented:
Finally; all green again 😅

neersighted reviewed (approved):
LGTM, looking forward to seeing more of this indirection and the big locks fall away.
